### ISSUE TITLE
Argo server did not recover from a connection drop from archive persistence layer (PostgreSQL)

### ISSUE THREAD

**Issue by kaancfidan** [argoproj/argo-workflows#7442]
The below error happens periodically in our system. We don't know what causes it or how to reproduce it, but I would like to at least have some pointers on how to investigate it further.

### Discussed in https://github.com/argoproj/argo-workflows/discussions/7300

<div type='discussions-op-text'>

<sup>Originally posted by **kaancfidan** November 30, 2021</sup>
I initially wanted to create an issue for this but I'm not sure how to reproduce it.

We have been using Argo Workflows in our AKS cluster on Azure for a year now and recently we have added the workflow archive by connecting the Argo server to a managed Azure Database for PostgreSQL. It had been running without any problems for ~2 weeks and this morning we have seen that the workflow APIs were returning Internal Server Error status code and when we checked the logs this was repeating:

> time="2021-11-30T07:29:35.369Z" level=error msg="finished unary call with code Unknown" error="write tcp 10.244.33.119:35646->10.10.0.7:5432: write: connection reset by peer" grpc.code=Unknown grpc.method=ListWorkflows grpc.service=workflow.WorkflowService grpc.start_time="2021-11-30T07:29:35Z" grpc.time_ms=20.129 span.kind=server system=grpc

Basically it was not able to reach the PostgreSQL instance.

PostgreSQL server in question was not down. The same instance is also used by other parts of our system and is accessed from the same cluster. So whatever network problem happened I assume it must have been transient. I restarted the argo-server deployment and the problem was fixed. Another weird thing is that I do not see any failed connection statistics on PostgreSQL side.

Connection drops are expected but I don't understand why Argo did not recover from it. Any ideas why this could occur?</div>

**Comment by kaancfidan** [comment]
I suspect it might be related to https://github.com/upper/db/issues/226.

**Comment by kaancfidan** [comment]
Any suggestions here? 

We are still experiencing outages and we had to basically put a liveness probe on the archive endpoint to restart argo-server whenever archive db connection fails.

**Comment by tooptoop4** [comment]
did u solve @kaancfidan

**Comment by kaancfidan** [comment]
We have put a liveness probe on the archived workflow API. I'm not sure if it's still necessary but it patched the issue for us at the time.
